Fix top position, height, and scrolling with modals

kristinalim opened the pull request:
#### What? Why?

- Closes #2201 
- Addresses unclickable button issue in #1766 

These changes will also affect other modals and tall dialogs (smaller screens with height less than 500px).

See the following video:

https://www.useloom.com/share/002a740fbee8436aadcfa41aaae4326a

Notice the following:

1. Vertical position of the modal is initially at 10px from the top of the screen. After scrolling down, however, the vertical position from the top of the screen varies.
2. When you are not currently at the top of the page, opening the modal usually scrolls the page up a little bit.
3. The modal sometimes goes past the bottom of the screen.

For large screens, the modal is actually styled to be `10% of height of the viewport` from the top of the screen. There was intention for the modals to start from the same position from the top.

Above behaviours were actually caused by three interconnected problems.

---

##### Fix 10% vertical position from the top of the screen

The modal JS library `angular-foundation` treats a % value for "top" CSS property as px value, so the intended 10% position from the top ends up being treated as 10px.

See details of the bug here: https://github.com/kristinalim/angular-foundation/pull/1

The code uses version `v0.8.0` of https://github.com/yalabot/angular-foundation. I planned to [submit a pull request](https://github.com/yalabot/angular-foundation/pull/319) addressing the issue, however, the library doesn't seem to be actively maintained anymore.

So, in my fork of the repository,  I created a `v0.8.1` release with the [bug fix](https://github.com/kristinalim/angular-foundation/pull/1), and also a [cherry-picked fix for unrelated failing tests](https://github.com/kristinalim/angular-foundation/pull/2).

I refrained from including other changes in the original repository's "master" branch (`v0.9.0-SNAPSHOT`) because I'm not sure if these are already stable enough.

(Note: The package is compatible with node `v0.10.48`. To compile, run `grunt`, and find the compiled and minified files in `dist/`. Following instructions in the `README`, these files were added to the ["gh-pages" branch](https://github.com/kristinalim/angular-foundation/tree/gh-pages).

##### Make sure the modal does not go past the bottom of the screen

When there is enough content in the modal, the height of the modal plus its offset from the top could exceed the height of the screen.

I changed the maximum height of the modal to 80% for large screens. Considering a top position of 10%, a max height of 80% renders a tall modal vertically centered, with 10% remaining space at the bottom.

##### Fix scrolling up when the modal opens
    
Occasionally, the page scrolls up while the modal is being opened. This was causing the final position of the modal to be at the wrong location relative to the top of the screen.

This was happening because of a race condition between the animation that slides the modal from above the screen to the middle, and `focus()` [which the modal does](https://github.com/yalabot/angular-foundation/blob/0.8.0/src/modal/modal.js#L109).

The final vertical position of the modal is 10% from the top, so the animation which translates the modal -25% vertically was starting -15% above the screen. The `focus()` then causes vertical scroll.

This lowers the starting point of the animation, so there will no longer be scrolling upon focus.

Additionally, the animation was updated to happen only on large screens. The CSS property "top" is 0 for smaller screens.

---

The following bug was also addressed.

##### Do not exceed screen height in smaller screens less than 500px tall

See the following video:

https://www.useloom.com/share/bf19aca8836142e1b321c484ecb2de25

#### What should we test?

For large screens:

1. Go the the product list of an enterprise.
2. Click on a product without scrolling.
3. Check that the modal starts at around ~10%~ 5% of the screen from the top.
4. Repeat Items 2 and 3 for a couple more products.
5. Scroll to the end of the list.
6. Remember which part of the page is currently visible.
7. Click on a product without scrolling.
    a. Check that the modal starts at around ~10%~ 5% of the screen from the top.
    b. Close the modal.
    c. Check that the visible part of the page is still the same, i.e. the page should not have scrolled up.
8. Repeat Item 7 for a couple more products.
7. Click on the a product with a long description.
8. Check that the modal starts at around ~10%~ 5% of the screen from the top, and also has a distance of ~10%~ 5% of the screen from the bottom.

For smaller screens with less than 500px height:

1. Go the the product list of an enterprise.
2. Scroll down to a product.
3. Remember which part of the page is currently visible.
4. Click on the product.
5. Check that the modal is in full screen mode, starting from the top of the screen and ending at the bottom.
6. Check that you can scroll all the way to the bottom of the content of the modal.
7. Close the modal.
8. Check that the visible part of the page is still the same, i.e. the page should not have scrolled up.

Other modals to check:

1. Login modal
2. Signup modal
3. Accept Cookies modal - Load homepage in incognito window.
4. Cookies Policy modal
5. Modal for editing product images, by clicking product image in admin product list.

#### Release notes

- Fix top position and page scrolling with modals
- Reduce maximum height of modals and dialogs in large screens
- Limit maximum height of modals and dialogs in small screens

Changelog Category: Fixed

luisramos0 commented on app/assets/stylesheets/darkswarm/modals.css.scss:
```diff
@@ -28,7 +28,7 @@ dialog, .reveal-modal {
   // Medium and up - when modal IS NOT full screen
   @media only screen and (min-width: 641px) {
     top: 10%;
-    max-height: 120%;
```
on my almost-merged cookies work I have changed this max-height to 100%. On the cookies policy page modal I want it to go to until the end of the screen: lots of content. Maybe we can set this to 95% to cover both cases?

kristinalim replied:
@luisramos0 Thanks for the heads-up! Seeing that your cookies banner has been merged into by now. I will rebase this branch upon master.

Hmm... If we retain the 10% position at the top but use a max height of 95%, we might end up with some hidden content again on some screen dimensions.

I see that your changes apply a CSS class "cookies-policy-modal" to the modal. I could try just adding custom styles that sets "top" to 2% and "max-height" to 96% (and animation to start at -2%) only for this particular modal. It would render similar to how it does in master now.

Are you okay with this?

luisramos0 replied:
yes, cookies policy modals can take it. I was just trying to avoid the !important:
  top: 2% !important;
  max-height: 96% !important;

Do you know how can we avoid these !importants?
One easy trick I used in the past is to have the generic/common css files (like modals.css in this case) be on the top of the all.css file. I am not sure how the stylesheets are loaded into all.css but it may well be alphabetically and because Modals comes after Cookies the styles on modals are read after and get the priority :-(

Any suggestions?

kristinalim replied:
Ah! Yes, you're right, the files would be loaded in alphabetical order, so `cookies_policy_modal.css.scss` would be loaded before `modals.scss`.

I'm seeing two options:

1. Use more specific selectors for the cookies policy modal, like this:

    ```css
    .cookies-policy-modal.reveal-modal { }
    ```

2. Load the CSS partial for the cookies policy modal last.

I think the second is more appropriate.

In my experience, page-specific CSS (I would categorize cookies policy modal CSS under this) are best to be loaded last, and could be loaded in any order among themselves because they are not reused.

So far, the file structure under `darkswarm/` is flat. I've been a fan of the [7-1 Pattern](http://sass-guidelin.es/#the-7-1-pattern), which guides the file structure of CSS partials and helps make their load order easier to manage. Following this, I would move the CSS specific to the cookies policy modal under `pages/`.

The following would then make sure that all the other CSS would still load in the way that they already do:

```scss
@import '*'; // This will not import subdirectories.
@import 'pages/*';
```

For a start, I would also move CSS for the other modals under `pages/`. The rest of the files, we could reorganize as we go along, or when we could properly audit. I'm also not sure yet how OFN instances make CSS customizations, so not sure how messy that would be.

What do you think?

luisramos0 replied:
perfect! 7.1 is fine. i am a fan of funcional organizations so id put cookiespolicy css next to the js files and htnl of that page. but rails is always organizing things by type(controllers/models/etc), which I dont like.
Anyway, what you described sounds awesome and if support that change!
i dont think instances have css customizations.
thanks!

kristinalim replied:
I found that the modal library does not allow us to easily customize the top position of just certain modals either. The faux modal it inserts to the DOM, from which it determines the top position of the real modal, does not apply any custom modal CSS class we use for the real modal.

To work around this limitation, I changed modals for large screens from 10% position from top and 80% max height, to 5% position from top and 90% max height. Before the recent changes, most modals were rendering at 10px top position (not the originally planned 10%), so changing this to 5% doesn't seem to be much of a compromise.

kristinalim replied:
I split `**/*` into a couple of lines, to allow us to start specifying the order in which we import the CSS partials. The files in `pages/*` are loaded last.

About the taller cookies policy modal, it seems that it is not easy to apply custom position and height to modals - this doesn't seem to be supported by the modal library.

To work around this limitation, I went ahead and made a general change for all modals in large screens to start from 5% and have 90% max height, instead of 10% and 80%. Before these changes, most modals were rendering at 10px top position, not the originally planned 10%. Changing this to 5% doesn't seem to be much of a compromise, and it works out nicely.

Let me know what you think.

luisramos0 commented on app/assets/stylesheets/darkswarm/modals.css.scss:
```diff
@@ -20,9 +20,10 @@ dialog, .reveal-modal {
   // Small - when modal IS full screen
   @media only screen and (max-width: 640px) {
     height: 500px;
```
Hey, now that we are on this one.

I see @HugsDaniel  change past Feb2018:
from
-    min-height: 100%; // This is needed to make the height not the height of whole content page
to
+    height: 500px;

Is there a reason why the modals (for example the login modal) should not take the full screen if there's stuff to be shown?
I'd go for max-height: 100% here and use the empty space below the 500px point.
What do you think?

kristinalim replied:
I'm not sure, but I'm thinking the `min-height: 100%` could have been made before the following was added:

```css
body.modal-open {
  overflow: hidden;                                                                                                                                                                                                                                                        
}
```

The comment would have been true - 100% height for the modal would have occupied the full length of the then scrollable `body`.

kristinalim replied:
@luisramos0 I removed `height: 500px` for small screens, to allow the modal to go past this height. There doesn't seem to be any issue from doing that.

luisramos0 commented on app/assets/javascripts/admin/all.js:
```diff
@@ -60,7 +60,7 @@
 //= require moment/pt-br.js
 //= require moment/sv.js
-//= require ../shared/mm-foundation-tpls-0.8.0.min.js
+//= require ../shared/mm-foundation-tpls-0.8.1.min.js
```
Looks good! I think we must copy/move this fork to the OFN org somewhere like http://github.com/openfoodfoundation/angular-foundation

mkllnk replied:
Good idea to fork it to our organisation. There also seems to be another fork that had a couple of useful commits: https://github.com/cwadrupldijjit/angular-foundation

We could also open our pull request against that repository and see if it's maintained. If not, it could be useful to cherry-pick those commits for our fork.

I'm not sure if @kristinalim has organisation permissions to create new repositories though. That may be something we have to do. But are you happy to open another pull request with the other repository first and see if that guy would like to be the new maintainer?

kristinalim replied:
I can reassign the repository to OFN, no problem. I wouldn't need prior permission, but OFN would need to approve the transfer of ownership within one day.

I actually submitted a pull request to the repository already: https://github.com/yalabot/angular-foundation/pull/319 (Oops! There's a merge conflict with `0.9.0-SNAPSHOT` there - I will fix that.) But there is [hint](https://github.com/yalabot/angular-foundation/issues/311) that the repository will no longer be maintained.

kristinalim replied:
Ah, I didn't realize that you meant I could try opening a pull request at https://github.com/cwadrupldijjit/angular-foundation instead. :slightly_smiling_face: I will do that. Thanks, @mkllnk!

kristinalim replied:
Here are the PRs to the original repository and another fork:

* https://github.com/yalabot/angular-foundation/pull/319
* https://github.com/cwadrupldijjit/angular-foundation/pull/1

I created a release [`0.9.0-20180826174721`](https://github.com/kristinalim/angular-foundation/releases/tag/0.9.0-20180826174721) which is based on the "master" branch in https://github.com/cwadrupldijjit/angular-foundation/ and cherry-picks the modal fixes. (The compiled JS is in the `gh-pages` branch, following the process for the `angular-foundation` project.) This is the version which this PR now uses.

The instructions about transferring repositories which I read must have been for non-organization Github accounts. I couldn't initiate a transfer to `openfoodfoundation`. Would a fork of my fork will be sufficient, while we're waiting to see if any of the PRs could get merged?

sigmundpetersen commented:
@luisramos0 please review the updates and move to Test Ready if good

kristinalim commented:
Can I get temporary access to create repositories on openfoodfoundation? Right now, I'm getting a permission error when I attempt to transfer the `angular-foundation` repository.

luisramos0 commented:
@sigmundpetersen we need to move the repo first.
@kristinalim I don't have that permission.

sauloperez commented:
try again now @kristinalim

kristinalim commented:
@sauloperez I still get this: "You don’t have the permission to create repositories on openfoodfoundation"

luisramos0 commented on app/assets/stylesheets/darkswarm/modals.css.scss:
```diff
@@ -20,7 +20,6 @@ dialog
 
   // Small - when modal IS full screen
   @media only screen and (max-width: 640px) {
-    height: 500px;
```
❤️

kristinalim commented:
I was already [able to transfer the repository](https://github.com/openfoodfoundation/angular-foundation) to the "openfoodfoundation" Github org. (Thank you, @sauloperez!)

I also added section "Other modals to check" in recommended testing notes above.

Moving this PR to "Test Ready".

mkllnk commented:
@kristinalim There is a conflict that needs resolving now, sorry.

kristinalim commented:
Thanks, @mkllnk! I rebased this and did some checks, and moved this back to "Test Ready".

sauloperez commented:
I'm very sorry @kristinalim. More conflicts :see_no_evil:

kristinalim commented:
Haha, @sauloperez. :laughing: I rebased, and also checked that changes of @luisramos0 in #2611 are still okay. Checks are now green, though the build on Semaphore failed a couple of times first.

luisramos0 commented:
staged here: https://staging.openfoodfrance.org/

sstead commented:
**Testing notes**

There were no issues with this PR. 

One thing that could be improved is the display of backend product image modal in a small screen view. Not crucial.

There's one thing I couldn't test... could @myriamboure or @RachL  please turn on cookies for the instance, and then check that the cookies modal displays correctly on big screens, small screens and mobile? Thanks. It's in Configuration > General Settings 
![image](https://user-images.githubusercontent.com/10843056/45194612-18c26900-b297-11e8-9c5b-707c614c2c14.png)

https://docs.google.com/document/d/1V41s7QatTIDCiss6L905tZdH6oyrPiZ8BNupUnO5Ksw/edit#

luisramos0 commented:
@sstead "There's one thing I couldn't test..." isuper admin is admin@openfoodfrance.org pwd spree123

RachL commented:
I've tested and added my screenshot to your document @sstead. However I agree that the requirement of issue #1766 is not met. Can we unlick this issue so it stays open and move the rest to ready to go? I've never done this so far. cc @luisramos0 @kristinalim

luisramos0 commented:
If you move this to ready to go, only the linked issue 2201 will be moved. 1766 will stay as it is I believe.

daniellemoorhead commented:
>If you move this to ready to go, only the linked issue 2201 will be moved. 1766 will stay as it is I believe.

Unfortunately/fortunately this isn't the case @luisramos0 so I've disconnected the PR. @RachL it would be worth putting a note on the original issue to say what is remaining to be done.

luisramos0 commented:
Are we moving this PR to ready to go and keeping the issue open? (that's what I understood)
Can I stage something else on staging-fr?

kristinalim commented:
@luisramos0 Yes, as far as I understand, we are in agreement that this PR is okay for the next release. #1766 in mobile is still better off with these changes.

I updated the PR description to specify that only the issue of the button in #1766 being unclickable is resolved by this PR. I also moved #1766 back to "Dev Ready". (I reopened it, so it was in "All the Things".)

luisramos0 commented:
awesome @kristinalim 
@RachL or @sstead can you pls confirm and move it to ready to go?

RachL commented:
@luisramos0 yes confirming and moving this along
